### PR TITLE
Feature: guess from audio file

### PR DESCRIPTION
**Pull request by dscottboggs**
All the pieces are in place but this doesn't yet work. See the test, I created a simple audio file of a piano *Am@4* and the notes that are guessed are *D♯3*, *C3*, and *G3*. I'm not too familiar with the math behind this whole process, I'm wondering if I've selected the wrong sample rate, or I'm doing the transcoding wrong somehow?

Any thoughts?

### CODE REVIEW

File: src/debug.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#[macro_export]
+macro_rules! debug {
+  ($fmt_str:literal $(, $args:expr)*) => {
+    #[cfg(feature = "debug")]
+    eprintln!($fmt_str, $(,$args)*);
```
Comment (twitchax) [inline src/debug.rs:5]: It might be a better option to add logging, and make this a `trace`?

File: src/bin.rs [hunk 2]
```diff
@@ -138,14 +159,28 @@ fn start(args: Args) -> Void {
...
+                analyze::preview_audio_file_segment(
+                    &source,
+                    start_time.as_ref(),
+                    end_time.as_ref(),
+                    ffmpeg_debug,
+                )?;
             }
-        },
+            let notes =
+                analyze::get_notes_from_audio_file(source, start_time, end_time, ffmpeg_debug)?;
```
Comment (twitchax) [inline src/bin.rs:181]: Nit: no need for break here.

**Review by twitchax — CHANGES_REQUESTED** [review]

**Comment by twitchax** [comment]
> All the pieces are in place but this doesn't yet work. See the test, I created a simple audio file of a piano _Am@4_ and the notes that are guessed are _D♯3_, _C3_, and _G3_. I'm not too familiar with the math behind this whole process, I'm wondering if I've selected the wrong sample rate, or I'm doing the transcoding wrong somehow?
> 
> Any thoughts?

Good question.  I can take a look at the mp3.  The sampling rate _should not matter_.

**Comment by twitchax** [comment]
Are you sure that the sample rate of your file is 44.1 kHz?

**Comment by dscottboggs** [comment]
I wasn't trying to assert that the sample rate of the input file is 44.1kHz, I was trying to convert any input to *be* 44.1kHz, so that we can calculate the length.

Based on your review, I think parsing the time at the CLI and then reconstructing the string before passing it to ffmpeg, which would negate the need to calculate the length from the count of samples received. I too suspected the sample rate being wrong may be what's causing the guesses to be wrong, so lets see if that fixes the problem before giving it too much more thought haha.

File: src/debug.rs (hunk 1 above)
Comment (dscottboggs) [reply src/debug.rs:5]: Sure, do you have a preferred [logging frontend crate](https://crates.io/crates/log#in-executables)?

File: src/analyze.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    file: impl AsRef<Path>,
+    start: Option<impl AsRef<str>>,
+    end: Option<impl AsRef<str>>,
+    ffmpeg_debug: bool,
+) -> Res<Vec<Note>> {
+    let mut cmd = Command::new("ffmpeg");
+    // pass ffmpeg the file to decode
+    cmd.arg("-i").arg(file.as_ref());
+    if let Some(start) = start {
+        // pass ffmpeg the start time
```
[the remark this answers is not in the document]
Comment (dscottboggs) [reply src/analyze.rs:40]: Sorry, I feel like I have a tendency to undercomment and overcompensated.

File: src/debug.rs (hunk 1 above)
Comment (twitchax) [reply src/debug.rs:5]: I normally just use `log`.  The consumer of the library can choose to use an actual logger to emit the logs to console / etc.

**Review by twitchax — CHANGES_REQUESTED** [review]

**Comment by dscottboggs** [comment]
> The consumer of the library can choose to use an actual logger to emit the logs to console / etc.

well, in this case we have both the library and the consumer. I was asking what you would prefer for the latter.

**Comment by dscottboggs** [comment]
Closing in favor of #3

**Pull request closed without merging**